PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "Defines.hpp"
+
+namespace Poly {
+	template<typename V>
+	class Optional {
+	public:
+		using ValueType = V;
+	public:
+		Optional() : initialized(false), dummy{} {}
+		Optional(const V&  value) : initialized(true), value_storage(value) {}
+		Optional(      V&& value) : initialized(true), value_storage(std::move(value)) {}
+		~Optional() { if (this->HasValue()) { this->Value().~V(); } }
+
+	public:
+		Optional(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				::new(&this->Value()) V(std::move(other.Value()));
+
+				other.Value().~V();
+				other.initialized = false;
+			} else {
+				this->initialized = false;
+			}
+		}
+		Optional& operator=(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				this->Value() = std::move(other.Value());
+
+				other.Value().~V();
+				other.initialized = false;
+			} else if (this->HasValue()) {
+				this->Value().~V();
+				this->initialized = false;
+			}
+			return *this;
+		}
+
+	public:
+		bool HasValue() const { return this->initialized; }
+		operator bool() const { return this->HasValue(); }
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&& Value() &&     { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return std::move(this->value_storage); }
+
+	public:
+		V TakeValue() {
+			ASSERTE(this->HasValue(), "Attempting to take a value from an empty optional");
+
+			V ret = std::move(this->value_storage);
+
+			this->initialized = false;
+			this->value_storage.~V();
+
+			return ret;
+		}
+
+	public:
+		V ValueOr(V&& fallback) const &  { if (this->HasValue()) { return this->Value(); } return fallback; } //todo(vuko): do not copy when possible
+		V ValueOr(V&& fallback) const && { if (this->HasValue()) { return std::move(this->Value()); } return fallback; }
+
+	private:
+		bool initialized;
+		union {
+			V value_storage;
+			char dummy;
+		};
+	};
+
+	template<typename V>
+	class Optional<V&> {
```

Comment:
It sure is! The specializations may look the same, but they're actually different underlying implementations with the same API.